### ISSUE TITLE
Global eyesight in case of block size 5

### ISSUE THREAD

**Issue by tamubun** [leela-zero/leela-zero#708]
In issue [#658](https://github.com/gcp/leela-zero/issues/658) I wrote spatially too separated points in a board has no related policy output values in case of small block sizes.

I had experiment to check this at some extreme positions: 
![boards](https://user-images.githubusercontent.com/5386688/35194470-6c1c137a-fef7-11e7-8a0f-65856a6b6b77.png)
Last moves are shown in above figure. Previous 7 moves are all far top right, bottom left, or bottom right corners to erase effects to the top left corner. The last white move in position A is critical for large black group. On the other hand in position B, the large group is not vulnerable.

Using training/tf/tfprocess.py (a little modified), and the visualization of outputs layers is
![output](https://user-images.githubusercontent.com/5386688/35194516-2f4d1c0e-fef8-11e7-9d89-09ae74588ab6.png)
I used the network c83e1b6e. In above figure, two planes (before flattening) of FC layer inputs is shown left, and final policy output is shown right among with the difference of outputs between A and B positions. The range means what value corresponds white most colour and red most colour.

The values at top left 5x5 region (A19-E15) in FC layers inputs of A, B positions are exactly same, and differences are 0.0. Final policy values at the same region are different between A, B positions by the effect of FC layer, but very small. That is only the global effect.

**Comment by FFLaguna** [comment]
Does the NN itself need to learn ladders? Can't the NN + MCTS eventually read ladders extremely quickly and easily?

I was watching the most recent six-block LZ weights (7fde81e8) play on Sabaki, and from the GTP output it had read a 14-move ladder fully in less than 2,172 playouts. (wG10 would have ended in a failed white ladder. bL7 would capture it at the end.)

At 4,396 playouts, LZ for some reason still preferred the failed ladder. But by 6,622 playouts LZ had found a move that was better than the failed ladder.

```
Playouts: 2172, Win:  7.79%, PV: G10 H10 G9 G8 H9 J9 H8 H7 J8 K8 J7 J6 K7 L7 G12
Playouts: 4396, Win:  7.70%, PV: G10 H10 G9 G8 H9 J9 H8 H7 J8 K8 J7 J6 K7 L7 H11
Playouts: 6622, Win:  7.67%, PV: G18 G10 H5 Q10 Q11 R9 S9 R11 S10 R12
Playouts: 8853, Win:  7.74%, PV: G18 G10 H5 Q10 Q11 R9 S9 R11 S10 R12 Q12 Q13
Playouts: 11060, Win:  7.66%, PV: G18 G10 H5 Q10 Q11 R9 S9 R11 S10 R12 P10 Q9 P11 O9
Playouts: 13272, Win:  7.68%, PV: G18 G10 H5 Q10 Q11 R9 S9 R11 S10 R12 P10 Q9 Q12 R13
Playouts: 15452, Win:  7.58%, PV: G18 G10 H5 Q10 Q11 R9 S9 R11 S10 R12 P10 Q9 P11 O9
Playouts: 17662, Win:  7.52%, PV: G18 G10 H5 Q10 Q11 R9 S9 R11 S10 R12 P10 Q9 P11 O9 Q13
Playouts: 19770, Win:  7.47%, PV: G18 G10 H5 Q10 Q11 R9 S9 R11 S10 R12 P10 Q9 P11 O9 Q13
```

**Comment by FFLaguna** [comment]
I was watching the most recent six-block LZ weights (7fde81e8 w/no random, no noise) play on Sabaki, and from the GTP output it had read a 14-move ladder fully in less than 2,172 playouts.

At 4,396 playouts, LZ for some reason still preferred the failed ladder. But by 6,622 playouts LZ had found a move that was better than the failed ladder.

Here is the relevant output below. wG10 ends in a failed white ladder, with bL7 capturing it at the end:

```
Playouts: 2172, Win:  7.79%, PV: G10 H10 G9 G8 H9 J9 H8 H7 J8 K8 J7 J6 K7 L7 G12
Playouts: 4396, Win:  7.70%, PV: G10 H10 G9 G8 H9 J9 H8 H7 J8 K8 J7 J6 K7 L7 H11
Playouts: 6622, Win:  7.67%, PV: G18 G10 H5 Q10 Q11 R9 S9 R11 S10 R12
Playouts: 8853, Win:  7.74%, PV: G18 G10 H5 Q10 Q11 R9 S9 R11 S10 R12 Q12 Q13
Playouts: 11060, Win:  7.66%, PV: G18 G10 H5 Q10 Q11 R9 S9 R11 S10 R12 P10 Q9 P11 O9
Playouts: 13272, Win:  7.68%, PV: G18 G10 H5 Q10 Q11 R9 S9 R11 S10 R12 P10 Q9 Q12 R13
Playouts: 15452, Win:  7.58%, PV: G18 G10 H5 Q10 Q11 R9 S9 R11 S10 R12 P10 Q9 P11 O9
Playouts: 17662, Win:  7.52%, PV: G18 G10 H5 Q10 Q11 R9 S9 R11 S10 R12 P10 Q9 P11 O9 Q13
Playouts: 19770, Win:  7.47%, PV: G18 G10 H5 Q10 Q11 R9 S9 R11 S10 R12 P10 Q9 P11 O9 Q13
```

Interestingly, as we can see, white LZ spent anywhere between 2200 and 4000 playouts search for better moves _**after**_ the ladder had already been determined to result in white failure.

Perhaps there is an issue in some way with the MCTS not demoting its most deeply read moves quickly enough?

**Comment by tamubun** [comment]
@FFLaguna Though the shapes of my examples is ladder, I'm not want to limit the discussion only about ladders, but my result show general LZ behavior for spatially spread large group.

**Comment by FFLaguna** [comment]
I think I see, so this post is _not_ an issue of reading sequences to an arbitrary depth.

Rather, it is about the five- or six-block NN not being able to determine the effect that a stone in the _bottom_ right corner has on stones in the _upper_ right corner.

Did I understand correctly?

**Comment by tamubun** [comment]
Yes, that is what I want to say. 

May be MCTS can read correctly, but I think this long range weak point affects MCTS too. For example, board position B of my example, the black big group is not in danger, but the bottom right corner pattern does not affect to top left corner because of the capacity of 5 blocks, and LZ policy output says top left corner should be examined in high priority on MCTS. If she could see bottom right corner, I think the priority to read top left become low, and LZ could read other region many times.

**Comment by evanroberts85** [comment]
I am not sure this demonstrates anything of value. The value of the extra stones in the bottom right to the top left is pretty much nothing unless you understand the ladder. If you don't understand the ladder, then correctly you would disregard the influence of the extra stones to the top-left area. So all this is really saying is that the c83e1b6e can not read the ladder properly

**Comment by evanroberts85** [comment]
Well yes, it thinks black will not be captured in both positions A and B. But we already know that Leela gets caught in ladders, but this does not in any way prove that it is due to the 5 block architecture.

**Comment by tamubun** [comment]
@evanroberts85 Maybe I should not choose ladder-like example. I only want to show long range effects, and ladder is not the matter.

**Comment by evanroberts85** [comment]
A better example would be where the choice of a joseki in one corner should
affect which direction to play when dealing with a 3-3 invasion.

**Comment by tamubun** [comment]
I don't try now, but from above result, for 3-3 invasion at C3 for example, when LZ think whether C4 or D3 is good, all stones on row 16,17,18,19 and column Q,R,S,T are 'almost' ignored by policy network if block size is 5. I wrote 'almost' because only full connected layer can affect long range.

This sates only on policy network, never confuse MCTS or value network.

**Comment by killerducky** [comment]
@tamubun nice visualizations! Can you make a different example where the stones that seem to be in atari are in a straight line? People are confused by the ladder shape. Or you could find some examples of LZ missing ataris in other situations near endgame. People love to post those pointing to how it's missing obvious ataris.

**Comment by tamubun** [comment]
I think 3-3 invasion example is interesting, so I want to try it tomorrow.

**Comment by tamubun** [comment]
Now I have checked other examples which are not misleading ladder-like shapes. The network is again c83e1b6e (64x5).
![board_l](https://user-images.githubusercontent.com/5386688/35211144-8c700220-ff98-11e7-9d5e-99ed52dfe605.png)
The last white move of position C breaks one eye at bottom right and requires black to urgent move at S17 to alive big stone group. In position D, the last white move is bad and black big group is alive without any cares. But to make these decisions, LZ must see large spanned area.

The values of output of policy layer, and input of FC layer are:
![output_l](https://user-images.githubusercontent.com/5386688/35211497-ff022d4e-ff99-11e7-9952-c0b41fd3f8ed.png)
From this visualization, we can see LZ do not distinguish top right corner of C and D.

There's more interesting result. I checked similar positions but with short length.
![board_s](https://user-images.githubusercontent.com/5386688/35211578-6cea1a1a-ff9a-11e7-82b9-0bb3d45baeca.png)
And the values are:
![output_s](https://user-images.githubusercontent.com/5386688/35211621-96f81cd0-ff9a-11e7-9451-c6ab54dc4c64.png)
The difference of bottom right corner between E and F is correctly propagates to up, and LZ properly distinguish to cases and she knows only E position needs care to move S6.

I have another experiment to check 3-3 invasion opening case as  evanroberts85 says, but to prepare figures takes a lot of times, so I'll divide posts.

**Comment by tamubun** [comment]
Another example of 3-3 invasion opening.
![joban](https://user-images.githubusercontent.com/5386688/35214224-caa09892-ffa3-11e7-81d4-89dc56916c93.png)
Position G is from game 3b743adb, and position H is modified by hand which move all stones at top right, bottom left, and bottom right to borders. In both positions, white's move is C17 (3-3 invasion). Usually strong black players read where to move considering all board around. But the following result shows LZ does not look all 4 columns from Q to T, and all 4 rows from 1 to 4.
![output](https://user-images.githubusercontent.com/5386688/35214707-64890588-ffa5-11e7-9e0a-7e4ebc531955.png)
One comment. The policy value difference G-H has some non-zero values on top left corner

**Comment by evanroberts85** [comment]
I really like example C&D with E&F, they are much more conclusive. The life and death shape is simple so there should be no problem Leela Zero solving them, except for the length of the group. This is unlike the ladder example as we know even AGZ took a lot of training before properly understanding ladders.

G&H are much less conclusive, the rule here is to generally block on your weaker side, but it is not clear which side is weaker as the bottom left and top right are not much different in both examples in terms of their influence on the top left. If I can not be sure to play, I would not expect LZ to have a strong opinion either and as you state G-H has a small non-zero value which sounds about right.

**Comment by tamubun** [comment]
@evanroberts85 While I was writing, I send my comment unfinished by mistake. On G & H, the values of D16 to A19 are exactly same beneath full connected layer. Though there are no nearby stones except D16, I think strong players will consider other corners more or less.

**Comment by tamubun** [comment]
Now I want to move up head part of E case.
-  When head territory is R7,S7 and T7, LZ knows S7 is urgent, though she likes Q16 or other.
-  When head territory is R8,S8 and T8, LZ likes S8 and T1 almost evenly.
-  When head territory is R9,S9 and T9, LZ likes T1 more than S8.

**Comment by wctgit** [comment]
_[Forgive me if I butcher the description of the architecture of the NN; I haven't yet read any of the relevant papers, so I'm just trying to paint a more or less vague picture of different scales. I hope the gist of what I mean gets across, even if I mess up on the actual details.]_

I think this kind of effect should not be so surprising. There are probably two main aspects which are limiting LZ's ability to perceive long-distance structures (and threats). 1) The relatively small size of the convolutional networks (3x3, 5x5, 6x6, whatever), but perhaps most importantly 2) The fact that LZ is still in the early stages of training.

But before I expand on those topics, @tamubun I have a couple question: You have shown very well that the policy output seems to be blind to long-distance structure on the immediate board position. But have you tested what LZ actually suggests as a PV for different depths of MCTS search? How deep is required before LZ notices the necessity to play the key move? Is it forever blind, or is there some reasonable depth at which it 'catches on' that the key move is important?

Suppose there _is_ some depth at which it generates PVs which start to prioritize the key move correctly. What does the policy output show for the board positions at those depths? Perhaps the value network's evaluation dominates at those depths, and the sheer loss of many stones outweighs the immediate ignorance of the strategic danger.

I hope my questions aren't coming across the wrong way. I'm just curious if this 'immediate long-distance blindness' is a fatal flaw for LZ, or if it's able to reach the right conclusion given enough time.

Back to the theoretical aspects: 

1) Convolutional NNs are inspired by the visual cortex of animal brains, especially the human visual cortex. Due to their biological and extremely parallel nature, the first layers of the visual cortex are much closer to 1-to-1 representations of the actual image on the retina than deeper layers. The next deeper layers begin to pick out small features like line segments, joints, etc. Deeper layers combine these features into shapes and whatnot. Eventually you can 'see' an entire visual scene.

But there are many layers, and they are massively interconnected. From first to second layer may be more akin to a 64x64 NN convolution than 3x3 or even 6x6. And there may be perhaps dozens, hundreds, or even thousands of layers deep (don't quote me on these numbers! :-) ) The point is simply that due to sheer massive parallelness, our brains are able to scale up to the complexity of the problem much more efficiently than even the most modern TPU-laden distributed computer architecture we could cobble together today. That may change one day, but for now we are very limited.

Obviously, the massively interconnected layers in the visual cortex are able to process massive amounts of visual data quite literally in the blink of an eye. And the many deep layers allow us to pick out extremely abstract features from our field of vision. So it's not difficult for us to immediately 'see' in an instant (i.e. without requiring any deep thought analogous to MCTS) that S17 needs to be played to make two eyes, or that the ladder works or doesn't work.

My point is simply that I think it will be a while before a non-super-computer project like this can afford the computational power to glance at a 19x19 board and immediately spot the crucial global move at the first level of policy network evaluation.

It would be awesome if we could afford 19x19 1x1 convolutions, connected to 18x18 2x2 convolutions, 17x17 3x3 convolutions, ..., all the way to a 1x1 19x19 fully connected layer. Or something along those lines. Our biological brains achieve something on the order of that magnitude, but our hardware and software aren't up to it yet. They remain a bit near-sighted; long-distance structure remains fuzzily out of focus to them.

Hopefully the MCTS part allows LZ to cover that weakness. It seems the example of AlphaGo Zero hints that it can be done. I'm optimistic.

Sorry for rambling a bit, but I guess my point is mainly that the brain is able to do this kind of stuff easily, but we're stuck with relatively slow hardware and limited memory/storage, so we have to try to come up with algorithms that bridge the gaps to achieve 'good enough' within a 'reasonable time'. It simply may not be possible for any current neural net to be able to detect those urgent global plays in a single evaluation of a policy network. But maybe it doesn't have to.

2) The second point is simply that maybe (1) isn't even the problem. I'm not 100% convinced that it is, although I'm maybe leaning like 90% that way. Maybe it's simply the case that LZ hasn't had enough time to **learn** to overcome her short-sighted blindness. But maybe she really can. Even if 6x6 with 64 filters will never be able to identify those long-distance problems immediately, maybe it **will** be possible with 6x6 and 128, or some other achievable scale -- but maybe it will take a lot more training before that blindspot becomes important enough to show up in the network weights. Maybe the low values and zero values that we see right now are simply because LZ is still relatively an infant at Go. Maybe in a year or so, it will start to play well enough that ladders and other long-distance structures suddenly start to become the crucial difference between winning and losing games, and suddenly the training data will start to reflect those kinds of games, and suddenly the NN will start to adapt its weights to take those kinds of positions into account.

The basic point here is simply: Maybe it's just too early for us to say for sure that it **can't** be done. Maybe LZ will surprise us soon and show that she **can** make sense out of these positions. I don't think that's terribly unreasonable; it could simply be a matter of time.

**Comment by tamubun** [comment]
@wctgit I'm also thinking current limited LZ is strong enough. I can't beat her at all since last year's end. But I think you do not look at one important aspect, which I realized by your comment.

Suppose LZ finds out a good move which is related long spanned distance by MCTS, no matter how short she can see. What is important is _the reinforced learning for policy network is constructed to predict this good move only from board position (without MCTS)_, but sadly LZ can't see far distance. This is like: You are seid repeatedly you should learn 'stop' when you see red signal at 100km away, and 'go' when you see green signal there. Current LZ may always be forced to learn these impossible demand.

**Comment by tamubun** [comment]
The pattern E is very interesting for me, and I played more.
![nakades](https://user-images.githubusercontent.com/5386688/35282431-d3f531fa-0098-11e8-9433-754cef27a53a.png)

**Comment by evanroberts85** [comment]
So the 6x128 network still struggles to read large groups, but does better than 6x64.

Hopefully we have a 10x128 network to test soon.

**Comment by godmoves** [comment]
Can you share the code you use. I am also very interested in this topic, maybe I can test the performance of the net that I have trained before:)

**Comment by wctgit** [comment]
_[again, apologies for my ignorance of the papers, my question may be naive.]_

@tamubun 

Are you certain that the training of the policy network is **100%** independent of the MCTS? Doesn't the training proceed from the self-play games, from which the next moves are chosen **with** the involvement of the MCTS? If that's as I imagine, then shouldn't the policy network be able to incorporate such long-distance information once the games become strong enough to involve such long-distance threats more frequently?

**Comment by tamubun** [comment]
@godmoves I forked LZ and put the code in [tamubun branch](https://github.com/tamubun/leela-zero/tree/tamubun).

I reused my old code, which uses [RocAlphaGo 315f89d](https://github.com/Rochester-NRT/RocAlphaGo/tree/315f89d8dbb40256462f3301a20cdf4dc7354ad2) with some modifications. I put these files in my fork, but if it is not bad manner, I should delete them.

**Comment by tamubun** [comment]
@wctgit Sorry for my poor English. I want to say that " the training of the policy network is dependent of the MCTS". As for AGZ paper, policy network is trained to predicate previous MCTS result.

**Comment by tamubun** [comment]
From comment [zediir](https://github.com/gcp/leela-zero/issues/591#issuecomment-359946111) I got 20x256 supervised network and again checked pattern E (and C).
![nakades_20](https://user-images.githubusercontent.com/5386688/35441161-bf51c138-02e5-11e8-9349-ba9bdf7fc0ad.png)
It seems the network can see long range distance, but unfortunately it likes bad position T1 even if the range is very short.

**Comment by zediir** [comment]
I'll upload a 256x20 net with ~500,000 more training steps in a couple of hours when I get home. I'm going to keep training it for a couple of more weeks and soon try lowering the learning rate. 

It is lot slower training 256x20 than 64x5 or 128x6 because I had to go down to size 64 minibatches to get them to fit in memory.

**Comment by tamubun** [comment]
@zediir From above experiment, I know 256x20 can see all board at once. Now I want to check the  network of this size understand ladder, and ladder break (without MCTS). Maybe 500,000 more training steps is enough, or may not enough.

**Comment by st90115** [comment]
if it can see all boards with the block size less than  20 ?

**Comment by billyswong** [comment]
20-blocks net being able to see whole board at once makes sense intuitively. Deepmind also chose 256x20 as demo.

What I wonder is the capability of small networks like 64x10 or 64x12. We have seen LeelaZero making "low level mistakes (in the eye of humans)" in reading life&death of large groups. So in the distant future if we want to enjoy a set of various gameplay strengths for various level of players, we'd better figure out a minimum block size and use less filters for weaker bots, as the gameplays will likely be more "balanced" to human styles.

**Comment by lightvector** [comment]
I know this isn't in the spirit of an AlphaGo Zero replication, but in case anyone is curious, things like this are improved a bit if you add a "chain pooling" layer. Using the following functions:
https://www.tensorflow.org/api_docs/python/tf/unsorted_segment_max
https://www.tensorflow.org/api_docs/python/tf/gather
it's possible to implement a layer that performs max-pooling across connected chains of stones (rather than pooling across 2x2 squares as you normally see in the image literature).

On my own neural nets, different in architecture from Leela Zero and not adhering to quite a spirit of zeroness, I recently started testing such a layer. The architecture is a bit different, but roughly it's comparable to a 5-block resnet. As expected, for a small problem, the neural net does fine:
![normal-close](https://user-images.githubusercontent.com/11942395/35466635-71ee98d8-02d3-11e8-912c-3fa48ff3a77c.png)

But if we extend the tail, then it has trouble:
![normal-far](https://user-images.githubusercontent.com/11942395/35466659-ba679650-02d3-11e8-81e4-8df287186d25.png)

Adding a block with a single chain-pooling layer significantly improves this:
![chainpool-far](https://user-images.githubusercontent.com/11942395/35466662-cf59b976-02d3-11e8-8683-75f6f608f691.png)

With the white group being one solid chain like this, it the neural net will suggest S19 regardless of how far away the eye being poked is, so clearly it's using the chain pooling layer to convey two-eye information. But it still helps even when there's a gap in the chain that interrupts the information flow from the pooling, presumably chain pooling still is able to carry the information enough distance so that the 3x3 convolutions can do the rest of the job:
![chainpool-fargap](https://user-images.githubusercontent.com/11942395/35466663-d270cbf4-02d3-11e8-873c-8db9fd2eea47.png)

If you make it even farther away while adding yet more gaps in the chain spaced out along the way, then of course the neural net no longer suggests S19.

Unfortunately, the chain pooling operation, at least with my implementation, is fairly expensive. I haven't done a detailed comparison, but eyeballing it, on my machine I'd guess it slows down a training step by perhaps five to ten times as much as a convolution layer would.

**Comment by killerducky** [comment]
@lightvector Oh you got it working, awesome!

**Comment by lightvector** [comment]
Yeah, I'll update my Github in a week or two with some of the summaries. Testing is slow, since these things take some time to train. Some of the results I've gotten with chain pooling are counterintuitive - due to the expense I tried first only making the layer very small (bottlenecked it down to 32 channels, compared to the rest of the neural net which as 192), but that actually made things worse and caused the neural net to underfit for some reason. Putting in two of them and making them less bottlenecked (64 channels) does improve the loss, but of course it slows down training a lot, so I might hold off on doing so for now. On the margin right now, per unit of computation cost it's not obviously better than just adding more convolution blocks.

**Comment by sethtroisi** [comment]
recent networks are much larger and this is >1 year old, closing

**Comment by willemhendriks** [comment]
I am very interested in getting the raw signals of internal layers.

(?) Is it possible to put the code somewhere?

I am asking this as I am busy porting the network to Keras, and need test cases, layer-by-layer, to verify my architecture & weights transfer. My output currenly makes no sense and don't know where to look.

**Comment by alreadydone** [comment]
Maybe simply print out (using `myprintf`) `conv_out` after each `winograd_convolve3` in CPUPipe.cpp? e.g. https://github.com/leela-zero/leela-zero/blob/e542dd0d74b68682f624efbe1413151499403f7a/src/CPUPipe.cpp#L399
And append `--cpu-only` to command when running leelaz. 

And I think you need to take care of this when porting:
https://github.com/leela-zero/leela-zero/blob/e542dd0d74b68682f624efbe1413151499403f7a/training/tf/tfprocess.py#L490-L496
https://github.com/leela-zero/leela-zero/blob/e542dd0d74b68682f624efbe1413151499403f7a/training/tf/tfprocess.py#L365-L373

**Comment by willemhendriks** [comment]
Yes, thx!  Ill try that - i also found
https://github.com/tamubun/leela-zero/tree/tamubun
- I've seen those bias/beta juggling going on indeed. I have a guess how it should be but really need test

**Comment by willemhendriks** [comment]
Was something changed from 6x128 to 10x128?
I am able to export 6x128 to Keras, but somehow 10x128 output is not exactly the same with my Keras model.

**Comment by alreadydone** [comment]
10x128 spanned from Mar 5 to Apr 4, 2018. SWA (stochastic weight averaging) was added in between. After SWA, "batchnorm betas and gammas" in the weight files are no longer always 0 and 1.

Did you try the 91st and 92nd networks?

**Comment by willemhendriks** [comment]
I did a little research and indeed found https://github.com/leela-zero/leela-zero/pull/1112

My little understanding is that is should be compatible but apparently it is not. 
(?) Are there extra weights added during this transition?

If not, probably I need to infuse the beta in the same way as is done like your previous post. My current implementation doesn't, as I read

>  If this paragraph does not make any sense to you, ignore its existence and just add the channel bias layer as you normally would, output will be correct.

But maybe this is not that true as I now see.

Also - this probably means I need a layer-by-layer validation after April 2018, as currently I use tamubun branch from jan 2018.

**Comment by willemhendriks** [comment]
( currently I can't download any weight / hosting problems i assume)

**Comment by alreadydone** [comment]
> (?) Are there extra weights added during this transition?

I don't think there are. But I guess #1112 make the old training/tf code invalid, including tamubun's code. On the other hand, I don't think the SWA implementation change the cpp inference code in src/ and the .txt weight file format at all.

> If this paragraph does not make any sense to you, ignore its existence and just add the channel bias layer as you normally would, output will be correct.

Where did you see this?

> I am able to export 6x128 to Keras, but somehow 10x128 output is not exactly the same with my Keras model.

So you get output from three sources: your Keras model, tamubun's Python code (both layer-by-layer), and the leelaz executable (only the final output). Am I right? Which one doesn't agree with the other two? Or all of them are different with each other?

> Also - this probably means I need a layer-by-layer validation after April 2018, as currently I use tamubun branch from jan 2018.

Any trouble inserting print statements in CPUPipe.cpp? After that if you open up leelaz and type `heatmap` then you should see the outputs of each layer (if it's a nonempty board you'd like to fix the symmetry).

**Comment by willemhendriks** [comment]
> Where did you see this?

That line about bias/beta weights is from leela-zero github README.md
https://github.com/leela-zero/leela-zero#weights-format

Maybe - Slight changes in infusing the beta - implied roundoff errors, building up layer-by-layer. Anyway, there is no other way than layer-by-layer comparisons. 

I have only tested my Keras output, layer-by-layer, with tamubun's code, not `CPUPipe.cpp`. I'm a little sC++ared, as it is probably >20 years since I code and compiled my own C code. ::, ->, auto, what is going on!? Seems is no choice, i just have to add those `myprintf()`'s.

Thanks a lot for helping me with this. Here is my repo so far.

https://gitlab.com/whendrik/leelazerotokeras

When its done I will write some text about the exact weights format and implementation.

**Comment by willemhendriks** [comment]
I have some progress - the `myprintf()` is working now. 
I also found out the big difference,
up to my 10x128 test cases, all biases vectors were [0.0, 0.0,...0.0].

Not sure yet what is wrong, my Keras implementation or tamubun old code with new weights format.

--- 
update:

Turns out my Keras implementation is the same as `CPUPipe.cpp`, at least till the last block of the tower. I still need to test the 2 heads, which are different functions.
